php-classes/Slate/CBL/StudentCompetency.php

```diff
@@ -53,16 +64,349 @@ class StudentCompetency extends \ActiveRecord
+            ]
+        );
+
+        // starting rating can not be calculated if each required skill has not been rated at least once
+        foreach ($validSkillIds as $skillId) {
+            if (empty($ratedSkills[$skillId])) {
+                return null;
+            }
+        }
+
+        // get the first rating (by date) for each skill
+        $demonstrationSkillRatings = DB::allValues(
+            'skillRatings',
+            '
+            SELECT ds.DemonstratedLevel as skillRatings
+        	  FROM `%1$s` ds
+    		  JOIN `%2$s` d
+    		    ON d.ID = ds.DemonstrationID
+    		 INNER JOIN (
+                    SELECT MIN(demo.Demonstrated) as Created, SkillID
+    		          FROM `%1$s` demoskill
+    		          JOIN `%2$s` demo
+    		            ON demo.ID = demoskill.DemonstrationID
+    		         WHERE TargetLevel = %3$u
+                       AND DemonstratedLevel > 0
+                       AND Override = false
+                       AND demoskill.SkillID IN (%4$s)
+                       AND demo.StudentID = %5$u
+                     GROUP BY SkillID
+    		  ) ds2
+    		    ON ds2.SkillID = ds.SkillID
+    		   AND ds2.Created = d.Demonstrated
+    		 WHERE ds.SkillID IN (%4$s)
+    		   AND ds.TargetLevel = %3$u
+    		   AND ds.DemonstratedLevel > 0
+    		   AND d.StudentID = %5$u
+    		   AND ds.Override = false
+            ',
+            [
+                DemonstrationSkill::$tableName,
+                Demonstration::$tableName,
+                $this->Level,
+                join(', ', $validSkillIds),
+                $this->StudentID
+            ]
+        );
+
+        return array_sum($demonstrationSkillRatings) / count($validSkills);
+    }
+
+    private $demonstrationData;
+    public function getDemonstrationData()
+    {
+        if ($this->demonstrationData !== null) {
+            return $this->demonstrationData;
+        }
+
+        try {
+            $data = DB::arrayTable(
+                'SkillID',
+                'SELECT '.
+                '    DemonstrationSkill.*, '.
+                '    Demonstration.Demonstrated AS DemonstrationDate '.
+                '  FROM `%s` DemonstrationSkill '.
+                '  JOIN (SELECT ID, Demonstrated FROM `%s` WHERE StudentID = %u) Demonstration '.
+                '    ON Demonstration.ID = DemonstrationSkill.DemonstrationID '.
+                ' WHERE DemonstrationSkill.SkillID IN (%s) '.
+                '   AND DemonstrationSkill.TargetLevel = %u '.
+                ' ORDER BY SkillID, DemonstrationDate, DemonstrationID',
+                [
+                    DemonstrationSkill::$tableName,
+                    Demonstration::$tableName,
+                    $this->StudentID,
+                    join(', ', $this->Competency->getSkillIds()),
+                    $this->Level
+                ]
+            );
+
+            return $this->demonstrationData = $data;
+        } catch (TableNotFoundException $e) {
+            return [];
```

Comment:
@nbey this code will result in the empty result of a table being missing not being cached, defeating the purpose of our strict null check to allow for empty arrays as legit values.

Use this pattern instead for all of these functions:

```php
if ($this->demonstrationData === null) {
    try {
        $this->demonstrationData = DB::arrayTable(
            // ...
        );
    } catch (TableNotFoundException $e) {
        $this->demonstrationData = [];
    }
}

return $this->demonstrationData;
```